### PR TITLE
integration: add the resolveUrl method

### LINKED ISSUE #4351: Location Entity in Azure DevOps
When trying to store a location entity in a repo in Azure DevOps to evaluate an organization structure I came across an issue with the handling of `targets`. The specification looks like the following:

```yaml
apiVersion: backstage.io/v1alpha1
kind: Location
metadata:
  name: test-organization
  description: A collection of users and groups used for evaluation
spec:
  targets:
    - ./department.yaml
```

Lets say the file is stored in the following location: 
`https://dev.azure.com/organization/project/_git/repository?path=%2Fcatalog-info.yaml`

## Expected Behavior

I expect the referenced files in `targets` to be handled correctly. That means that the `department.yaml` should be fetched using the following URL:
`https://dev.azure.com/organization/project/_git/repository?path=%2Fdepartment.yaml`

## Current Behavior

Right now the URL used when fetching that file becomes the following:
`https://dev.azure.com/organization/project/_git/department.yaml`

The logic for this is here as far as I can see: https://github.com/backstage/backstage/blob/bc18571b7a742863a770b2a54e785d6bbef7e184/plugins/catalog-backend/src/ingestion/processors/LocationEntityProcessor.ts#L22-L34

I guess this works for GitHub and perhaps others as well but for Azure DevOps this needs some extra handling.

## Possible Solution

I guess it would be good to have some central function to reference other files in a repo since it can be different depending on the provider. This would not only be useful for this use-case but I can see us using it in the future when we want to look up things in `package.json` or `build.gradle` files for example.

Maybe this already exist and I'm just not aware of it. I checked the AzureUrlReader but that doesn't seem to support referencing another file in the repo. 

I guess my question is if there already is some functionality like this and if not what would be the best way forward for adding that?

### PR DESCRIPTION
**Pull request by freben**
Fixes #4351

### CODE REVIEW

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4353-ayy8hl2sbvh2ugsqbgndx6lqmuzz7hr8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/601923ce353bbbd7e69d3cd3

File: packages/integration/src/azure/AzureIntegration.ts [hunk 1]
```diff
@@ -42,4 +42,32 @@ export class AzureIntegration implements ScmIntegration {
   get config(): AzureIntegrationConfig {
     return this.integrationConfig;
   }
+
+  /*
+   * Azure repo URLs on the form with a `path` query param are treated specially.
+   *
+   * Example base URL: https://dev.azure.com/organization/project/_git/repository?path=%2Fcatalog-info.yaml
+   */
+  resolveUrl(url: string, base: string): string {
```
Comment (Rugvip) [inline packages/integration/src/azure/AzureIntegration.ts:51]: Wondering if it makes more sense to make this `resolveUrl(base: string, relative: string)`? Feeling a bit more intuitive since that's how `path.resolve` works. Also would allow us to make the relative bit a vararg.

File: packages/integration/src/azure/AzureIntegration.test.ts [hunk 2]
```diff
@@ -41,4 +41,24 @@ describe('AzureIntegration', () => {
...
+
+    expect(
+      integration.resolveUrl(
+        './a.yaml',
+        'https://dev.azure.com/organization/project/_git/repository?path=%2Fcatalog-info.yaml',
+      ),
+    ).toBe(
+      'https://dev.azure.com/organization/project/_git/repository?path=%2Fa.yaml',
+    );
+
```
Comment (Rugvip) [inline packages/integration/src/azure/AzureIntegration.test.ts:56]: Needs a test for

```ts
integration.resolveUrl('./a.yaml', 'https://dev.azure.com/organization/project/_git/repository')
```

File: packages/integration/src/ScmIntegrations.test.ts [hunk 3]
```diff
@@ -73,4 +73,16 @@ describe('ScmIntegrations', () => {
     expect(i.byHost('github.local')).toBe(github);
     expect(i.byHost('gitlab.local')).toBe(gitlab);
   });
+
+  it('can resolveUrl using fallback', () => {
```
Comment (Rugvip) [inline packages/integration/src/ScmIntegrations.test.ts:77]: Think we should make this work as one might expect?

```ts
i.resolveUrl('./catalog-info.yaml', 'https://github.com/backstage/backstage/tree/master')
```

File: packages/integration/src/azure/AzureIntegration.ts (hunk 1 above)
Comment (freben) [reply packages/integration/src/azure/AzureIntegration.ts:51]: On the other hand this is exactly the URL constructor names and order :) which is the one I think makes more sense to mimic. This one is also meant to be specialised for that two-arg use case, so probably no reason for a vararg

File: packages/integration/src/ScmIntegrations.test.ts (hunk 3 above)
Comment (freben) [reply packages/integration/src/ScmIntegrations.test.ts:77]: Hm to get that working we should override this method in the github integration. Separate PR?

File: packages/integration/src/azure/AzureIntegration.ts (hunk 1 above)
Comment (freben) [reply packages/integration/src/azure/AzureIntegration.ts:51]: Hey you know what. This is exactly the kind of method that should take a single object arg, way clearer. `{ url, base }`

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4353-ayy8hl2sbvh2ugsqbgndx6lqmuzz7hr8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/601923ce353bbbd7e69d3cd3

**Comment by mfrinnstrom** [comment]
@freben this was fast. Thanks a lot! I just tested it against our Azure DevOps repo and it works as expected 👍

File: packages/integration/src/azure/AzureIntegration.test.ts (hunk 2 above)
Comment (freben) [reply packages/integration/src/azure/AzureIntegration.test.ts:60]: added

File: packages/integration/src/azure/AzureIntegration.test.ts [hunk 4]
```diff
@@ -41,4 +41,43 @@ describe('AzureIntegration', () => {
...
+      const integration = new AzureIntegration({
+        host: 'dev.azure.com',
+      } as any);
+
+      expect(
+        integration.resolveUrl({
+          url: './a.yaml',
+          base: 'https://dev.azure.com/organization/project/_git/repository',
+        }),
+      ).toBe('https://dev.azure.com/organization/project/_git/a.yaml');
```
Comment (Rugvip) [inline packages/integration/src/azure/AzureIntegration.test.ts:80]: Would expect this to be `https://dev.azure.com/organization/project/_git/repository?path=%2Fa.yaml` right?

File: packages/integration/src/ScmIntegrations.test.ts (hunk 3 above)
Comment (Rugvip) [reply packages/integration/src/ScmIntegrations.test.ts:77]: Yeah could do it separately if you want, I think we can say that the current behaviour is borked enough that I don't think we can expect any broad usage that would rely on it

File: packages/integration/src/azure/AzureIntegration.test.ts (hunk 4 above)
Comment (freben) [reply packages/integration/src/azure/AzureIntegration.test.ts:80]: Good point - that's really a corner case I hadn't thought of. If you are at the root of the repo, doesn't it have a path of just `%2F`?

Comment (freben) [reply packages/integration/src/azure/AzureIntegration.test.ts:80]: It'll be a bit annoying to have to deduce and special treat the exact repo root - for all versions of azure - maybe git-url-parse can help somehoe

Comment (Rugvip) [reply packages/integration/src/azure/AzureIntegration.test.ts:80]: Nah just not param. Maybe we can just throw if `path` is not present for now? It's similar to the GitHub example, except that that generally works a bit weird with folders too

Comment (freben) [reply packages/integration/src/azure/AzureIntegration.test.ts]: Updated! Did a middle-way thing that'll work for both cases.

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4353-ayy8hl2sbvh2ugsqbgndx6lqmuzz7hr8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/601923ce353bbbd7e69d3cd3

**Review by Rugvip — APPROVED** [review]
:shipit: 👍
